pkg/envoy/rbac/policy.go

```diff
@@ -1,136 +1,84 @@
-		if len(permissionRuleList.AndRules) != 0 && len(permissionRuleList.OrRules) != 0 {
-			return nil, errors.New("Permission rule cannot have both AND & OR rules at the same time")
-		}
-
-		var currentPermission *xds_rbac.Permission
-
-		switch {
-		case len(permissionRuleList.AndRules) != 0:
-			// Combine all the AND rules for this Permission rule with AND semantics
-			var andPermissionRules []*xds_rbac.Permission
-			for _, andPermissionRule := range permissionRuleList.AndRules {
-				// Fill in the destination port permission
-				if andPermissionRule.Attribute == DestinationPort {
-					port, err := strconv.ParseUint(andPermissionRule.Value, 10, 32)
-					if err != nil {
-						return nil, errors.Errorf("Error parsing destination port value %s", andPermissionRule.Value)
-					}
-					portPermission := GetDestinationPortPermission(uint32(port))
-					andPermissionRules = append(andPermissionRules, portPermission)
-				}
-			}
-			currentPermission = andPermissions(andPermissionRules)
-
-		case len(permissionRuleList.OrRules) != 0:
-			// Combine all the OR rules for this Permission rule with OR semantics
-			var orPermissionRules []*xds_rbac.Permission
-			for _, orPermissionRule := range permissionRuleList.OrRules {
-				// Fill in the destination port permission
-				if orPermissionRule.Attribute == DestinationPort {
-					port, err := strconv.ParseUint(orPermissionRule.Value, 10, 32)
-					if err != nil {
-						return nil, errors.Errorf("Error parsing destination port value %s", orPermissionRule.Value)
-					}
-					portPermission := GetDestinationPortPermission(uint32(port))
-					orPermissionRules = append(orPermissionRules, portPermission)
-				}
-			}
-			currentPermission = orPermissions(orPermissionRules)
-
-		default:
-			// Neither AND/OR rules set, set permission to Any
-			currentPermission = getAnyPermission()
-		}
-
-		finalPermissions = append(finalPermissions, currentPermission)
+	if p.applyPermissionsAsAnd {
+		policy.Permissions = []*xds_rbac.Permission{andPermission(permissions)}
+	} else {
+		policy.Permissions = permissions
 	}
-	if len(p.Permissions) == 0 {
-		// No permissions specified for this policy, allow ANY
-		finalPermissions = append(finalPermissions, getAnyPermission())
+
+	return policy
+}
+
+// UseAndForPermissions will apply all permissions with AND semantics.
+func (p *PolicyBuilder) UseAndForPermissions(val bool) {
+	p.applyPermissionsAsAnd = val
+}
+
+// AddPrincipal adds a principal to the list of allowed principals
+func (p *PolicyBuilder) AddPrincipal(principal string) {
+	// We need this extra defense in depth because it is currently possible to configure a wildcard principal
+	// in addition to specific principals. We should instead not allow this, preventing this misconfiguration from
+	// happening through validating webhooks.
+	if principal == "*" {
+		p.AllowAnyPrincipal()
 	}
+	if !p.allowAllPrincipals {
+		// TODO(4754) return the trust domain of the first MRC, ***while grabbing the lock***
```

Comment:
I don't understand this comment. 
1. Why does this code have anything to do with the trust domain, which should be a part of the opaque principal passed to the builder
2. Expand MRC